src/docfx/validation/JsonSchemaValidatorExtension.cs

```diff
@@ -46,5 +56,202 @@ public bool IsEnable(FilePath filePath, CustomRule customRule, string? moniker =
+
+            var message = error.Message;
+
+            if (!string.IsNullOrEmpty(customRule.Message))
+            {
+                try
+                {
+                    message = string.Format(customRule.Message, error.MessageArguments);
+                }
+                catch (FormatException)
+                {
+                    message += "ERROR: custom message format is invalid, e.g., too many parameters {n}.";
+                }
+            }
+
+            message = string.IsNullOrEmpty(customRule.AdditionalMessage) ?
+                message : $"{message}{(message.EndsWith('.') ? "" : ".")} {customRule.AdditionalMessage}";
+
+            return new Error(
+                level,
+                string.IsNullOrEmpty(customRule.Code) ? error.Code : customRule.Code,
+                message,
+                error.MessageArguments,
+                error.Source,
+                error.PropertyPath,
+                error.OriginalPath,
+                customRule.PullRequestOnly);
+        }
+
+        private bool TryGetCustomRule(
+            Error error,
+            Dictionary<string, List<SourceInfo<CustomRule>>> allCustomRules,
+            out SourceInfo<CustomRule>? customRule)
+        {
+            if (allCustomRules.TryGetValue(error.Code, out var customRules))
+            {
+                foreach (var rule in customRules)
+                {
+                    // todo need confirm override order
+                    var r = rule.Value;
+                    if (r.PropertyPath != null)
+                    {
+                        // compare with code + propertyPath + contentType
+                        var source = error.Source?.File;
+                        var pageType = source != null ? _documentProvider.GetPageType(source) : null;
+                        if (r.PropertyPath.Equals(error.PropertyPath) && r.ContentTypes.Contains(pageType))
+                        {
+                            customRule = rule;
+                            return true;
+                        }
+                        continue;
+                    }
+                    else
+                    {
+                        customRule = rule; // system error
+                        return true;
+                    }
+                }
+            }
+            customRule = null;
+            return false;
+        }
+
+        private Dictionary<string, List<SourceInfo<CustomRule>>> MergeCustomRules(
+            Dictionary<string, ValidationRules>? contentValidationRules,
+            Dictionary<string, ValidationRules>? buildValidationRules)
+        {
+            var customRules = _config != null ?
+                _config.Rules.ToDictionary(
+                    item => item.Key,
+                    item => new List<SourceInfo<CustomRule>> { item.Value })
+                :
+                new Dictionary<string, List<SourceInfo<CustomRule>>>();
+
+            if (contentValidationRules != null)
```

Comment:
Is there a special reason why the behavior differs between `contentValidationRules` and `buildValidationRules`?

Reply:
the code get `contentValidationRules` and add some patch if it is pr-only. For buildRules, we will load all rules and apply override.